pkg/backend/snapshot.go

```diff
@@ -168,11 +168,21 @@ type sameSnapshotMutation struct {
 // step that forces us to write the checkpoint. If no such difference exists, the checkpoint write that corresponds to
 // this step can be elided.
 func (ssm *sameSnapshotMutation) mustWrite(old, new *resource.State) bool {
```

Comment:
@pgavlin Is there a lot of benefit in having this function?  It seems to be an optimization to avoid work (which i'm not opposed to).  However, it does add this sort of extra complexity that then has to be maintained.  Would it be problematic to just remove this funciton and always write the new checkpoint file?

Reply:
> Would it be problematic to just remove this funciton and always write the new checkpoint file?

I suspect that one of the reasons we do this is to reduce the amount of traffic that we send to the service in the "no-op" update case.  Writing a checkpoint can end up being an expensive operation (since it requires a call to the pulumi service) and we'd like to elide these as much as possible.

Reply:
@ellismg is correct. This function was added to reduce traffic in common cases like no-op updates.

Reply:
(see https://github.com/pulumi/pulumi/pull/1798 for more details)

Reply:
gotcha.  Would it possible to change the approach here.  Namely: compare the actual 'updated' checkpoint to the 'last' checkpoint and only upload if it changed.  In other words, flip things around so instead of having to do all these checks (and having to know this needs to be updated), we instead just always check if the update is necessary with a full diff comparison, and just upload if it actually changed.

Note: this can be done at a later time.  i'm just always wary of this sort of code since it seems easy to miss something.

Reply:
> we instead just always check if the update is necessary with a full diff comparison, and just upload if it actually changed.

My gut says that this is not possible, as this function may return `false` to indicate that the checkpoint need not be written if a resource has changes that are not semantically meaningful. I don't see how a full diff could do the same.

Reply:
@CyrusNajmabadi @pgavlin If you believe we should do anything here - feel free to open a new issue to track.